### ISSUE TITLE
Tracking issue for `#![register_tool]`

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#66079]
A part of https://github.com/rust-lang/rust/issues/44690.

Some tools (`rustfmt` and `clippy`) used in tool attributes are hardcoded in the compiler.
We need some way to introduce them without hardcoding as well.

https://github.com/rust-lang/rust/pull/66070 introduced a way to do it with a crate level attribute:
```rust
#![register_tool(my_tool)]

#[my_tool::anything] // OK
fn main() {}
```
The previous attempt to introduce them through command line (https://github.com/rust-lang/rust/pull/57921) met some resistance.

This probably needs to go through an RFC before stabilization.

**Comment by SimonSapin** [comment]
Is the tool’s name an arbitrary identifier, that doesn’t necessarily map for example to any crate name?

**Comment by petrochenkov** [comment]
@SimonSapin 
Yes, the tool name is arbitrary.

**Comment by jyn514** [comment]
This appears to only work for attributes, not lints.
```rust
#![feature(register_tool)]
#![register_tool(xyz)]
#![warn(xyz::my_lint)]
```
```
$ rustc unknown-lint.rs  --crate-type lib
error[E0710]: an unknown tool name found in scoped lint: `xyz::my_lint`
 --> unknown-lint.rs:3:9
  |
3 | #![warn(xyz::my_lint)]
  |         ^^^
```

**Comment by petrochenkov** [comment]
AFAIK, registering a lint tool is useless without also registering actual lints, which are registered with `rustc_interface` methods (together with their tools), so `register_tool` may only affect diagnostic wording in this case.

**Comment by jyn514** [comment]
> AFAIK, registering a lint tool is useless without also registering actual lints, which are registered with rustc_interface methods (together with their tools), so register_tool may only affect diagnostic wording in this case.

Then there is no way at all to add tool lints, because `rustc_lint` will give an error if it doesn't recognize the tool, even if there is a lint registered: https://github.com/rust-lang/rust/blob/edeee915b1c52f97411e57ef6b1a8bd46548a37a/compiler/rustc_lint/src/levels.rs#L307-L317
https://github.com/rust-lang/rust/blob/edeee915b1c52f97411e57ef6b1a8bd46548a37a/compiler/rustc_ast/src/attr/mod.rs#L36-L38

**Comment by SimonSapin** [comment]
For what it’s worth this is how Servo uses this feature:

https://github.com/servo/servo/blob/92d6c89a8f3ac3ace5dbf65871dfb7c188ce3f95/components/script_plugins/lib.rs
https://github.com/servo/servo/blob/92d6c89a8f3ac3ace5dbf65871dfb7c188ce3f95/components/script/lib.rs#L10-L17

**Comment by jyn514** [comment]
@SimonSapin that looks like it only uses a tool attribute and not a lint. `unrooted_must_root` isn't a 'tool' lint in the sense of this issue because it doesn't use a tool prefix (`servo::unrooted_musl_root` or something).

**Comment by SimonSapin** [comment]
Here is an example of attributed used by the lint: https://github.com/servo/servo/blob/3f7697690aabd2d8c31bc880fcae21250244219a/components/script/dom/raredata.rs#L21

**Comment by jyn514** [comment]
Yes, that is an attribute and not a lint.

**Comment by jyn514** [comment]
For an example of someone wanting to use tool lints (not just attributes), see https://internals.rust-lang.org/t/pre-rfc-implementation-defined-lints/13870.

**Comment by SimonSapin** [comment]
There is also a custom lint involved, and it looks like we can `allow` it without namespacing: https://github.com/servo/servo/blob/3f7697690aabd2d8c31bc880fcae21250244219a/components/script/dom/bindings/root.rs#L48

**Comment by jyn514** [comment]
We cleared up the confusion offline - plugins adding their own lints is already supported today; what's not supported is lints in a custom *namespace*, like `#[allow(servo::unrooted_musl_root)]`.

**Comment by petrochenkov** [comment]
>Then there is no way at all to add tool lints, because `rustc_lint` will give an error if it doesn't recognize the tool, even if there is a lint registered

I don't think this is correct behavior, `rustc_lint` should not prevent registration and use of such lints.
Lint registration and use have no relation to name resolution, and are performed through their own set of compiler interfaces, I don't think `#![register_tool]` should have any effect on them.

**Comment by petrochenkov** [comment]
Ah, wait, I'm wrong.
If the compiler encounters an attribute like `#[allow(clippy::foo)]` in a regular non-clippy run, then it must ignore it instead of reporting an "unknown lint" error if the tool is known, that's the whole point of tool lints.
Then yes, `#![register_tool]` should have an effect.

**Comment by jyn514** [comment]
> If the compiler encounters an attribute like #[allow(clippy::foo)] in a regular non-clippy run, then it must ignore it instead of reporting an "unknown lint" error if the tool is known, that's the whole point of tool lints.
Then yes, #![register_tool] should have an effect.

Do you have suggestions for implementing this? I found https://doc.rust-lang.org/nightly/nightly-rustc/rustc_resolve/struct.Resolver.html#structfield.registered_tools, but rustc_lint doesn't have access to a resolver (right now it doesn't even depend on `rustc_resolve`). I thought about moving it to the LintStore, but then rustc_resolve doesn't have access to a lint store.

**Comment by petrochenkov** [comment]
@jyn514 
I think you can simply go through crate attributes and collect the registered tools again, that's only few lines of code (`fn registered_idents` minus error reporting) and have zero effect on performance.

**Comment by Oliboy50** [comment]
👍 I think that this kind of tool could help to solve [such issue](https://github.com/awslabs/aws-sdk-rust/issues/156)

**Comment by smoelius** [comment]
Are there any outstanding issues regarding this feature?

What would it take to get it into the next stable Rust?

**Comment by petrochenkov** [comment]
I'm not aware of any issues with this feature, someone needs to prepare a stabilization report and submit a stabilization PR including it.

**Comment by smoelius** [comment]
Thanks for your response, @petrochenkov.

> someone needs to prepare a stabilization report and submit a stabilization PR including it.

Anyone can prepare those, e.g., me?

**Comment by jyn514** [comment]
@smoelius yes, you can make a stabilization report :) there are instructions at https://rustc-dev-guide.rust-lang.org/stabilization_guide.html

**Comment by smoelius** [comment]
I seem to have run into a small problem. I am getting `add #![register_tool(name)] to the crate root` errors when `#![register_tool(name)]` is already in the crate root.

It appears that when [`LintLevelsBuilder::new`](https://github.com/rust-lang/rust/blob/cfff31bc833070a00578bd6178160aeed56f28ba/compiler/rustc_lint/src/levels.rs#L62-L67) is called in an early context (specifically, [here](https://github.com/rust-lang/rust/blob/cfff31bc833070a00578bd6178160aeed56f28ba/compiler/rustc_lint/src/context.rs#L815)), `crate_attrs` can sometimes be empty. For my particular case, `krate.span` refers to a source file for a module that is not the crate's root module.

One way to fix this problem might be to ignore the error in early passes, and defer the check to late passes only.

To this end: is it possible to tell from the session whether the current pass is early or late? (I couldn't figure this out.)

If not, then perhaps an `error_on_unknown_tools` flag could be added to `new`, similar to the `warn_about_weird_lints` flag that it has now.

Either way, is deferring the check the "right" way to fix this problem?

You can reproduce the problem as follows:
```sh
git clone https://github.com/trailofbits/dylint -b register_tool
cd dylint
cargo +nightly build -p dylint
grep 'register_tool' dylint/src/lib.rs
```

**Comment by smoelius** [comment]
I am now more of the opinion that we should add an `error_on_unknown_tools` flag to `LintLevelsBuilder::new` like I described above.

Here is what I recently observed.

It happens that crates are lowered to HIR before the post-expansion early lints are run: https://github.com/rust-lang/rust/blob/15d9ba0133ce0b35348e1c8367afe00aec841ffa/compiler/rustc_interface/src/passes.rs#L459-L476

Lowering collects all of the attributes together for a crate. So what one could do is pass the collected attributes down through  `check_ast_crate` to `LintLevelsBuilder::new`.

This would solve the problem for post-expansion early lints, bot not pre-expansion early lints, i.e., [here](https://github.com/rust-lang/rust/blob/15d9ba0133ce0b35348e1c8367afe00aec841ffa/compiler/rustc_interface/src/passes.rs#L246-L253). So one would still need a flag to say: don't emit the "unknown tool" error during pre-expansion passes.

I can't see a way to collect all crate attributes together for a pre-expansion pass. So it would seem that one can push the earliness of this check only so far.

In my opinion, it is not worth jump through hoops to perform this check early. And I would consider "requiring that lowering be done before post-expansion early lint passes" to be jumping through such a hoop. (Mind you, there may be other approaches, but still.) Rather, I think it sufficient that the check just be performed before compilation terminates, e.g., during a late pass.

So, should I just submit a PR to add an `error_on_unknown_tools` flag to `LintLevelsBuilder::new`? What do you think? @petrhosek @jyn514

**Comment by jyn514** [comment]
@smoelius that seems fine :) if someone has concerns they can mention it in review, it doesn't prevent you from opening a PR.

**Comment by smoelius** [comment]
Thanks, @jyn514. I submitted the PR.

**Comment by jyn514** [comment]
> @smoelius yes, you can make a stabilization report :) there are instructions at https://rustc-dev-guide.rust-lang.org/stabilization_guide.html

Did you get a chance to follow up on this? :) I see you submitted a PR fixing a bug in the lint machinery, I'm not aware of any other bugs right now.

**Comment by smoelius** [comment]
Sorry, @jyn514, I've let this slip.

Are you asking because you wanted to make the report? If not, I can make it a priority to get this done.

**Comment by jyn514** [comment]
I'm asking because I'd love this to stabilize, but don't have the time to follow up on it myself. (I also got here from https://github.com/trailofbits/dylint/issues/28 just now :grin:)

> If not, I can make it a priority to get this done.

That would be great! :heart:

**Comment by smoelius** [comment]
# Stabilization Request

## Summary

This crate level attribute registers a tool with `rustc`. Such a registration tells `rustc` that attributes and lints prefixed with the tool's name can be ignored, as they are handled by a separate entity.

Example:

```rust
#![register_tool(my_tool)]

#[my_tool::my_attribute] // No "unknown attribute" error
fn foo() {}

#[deny(my_tool::my_lint)] // No "unknown lint" warning
fn bar() {}
```

Here is a [real-world use case](https://github.com/trailofbits/dylint/commit/059db43cec6ea5cc7957ce5424e8879c1b561b4a).

## Tests

-   [`register_tool` works independently of the deprecated `plugin` feature](https://github.com/rust-lang/rust/blob/master/src/test/ui-fulldeps/issue-40001.rs)
-   [`register_tool` properly handles attribute arguments](https://github.com/rust-lang/rust/blob/master/src/test/ui/attributes/register-attr-tool-fail.rs)
-   [Renaming of imports does not affect `register_tool`](https://github.com/rust-lang/rust/blob/master/src/test/ui/attributes/register-attr-tool-import.rs)
-   [`register_tool` and `register_attr` use distinct namespaces](https://github.com/rust-lang/rust/blob/master/src/test/ui/attributes/register-attr-tool-prelude.rs)
-   [`register_tool` must be used as an inner attribute](https://github.com/rust-lang/rust/blob/master/src/test/ui/attributes/register-attr-tool-unused.rs)
-   [`register_tool` works with `cfg_attr`](https://github.com/rust-lang/rust/blob/master/src/test/ui/attributes/register-attr-tool.rs)
-   [Tool lint levels can be set from the command line](https://github.com/rust-lang/rust/blob/master/src/test/ui/lint/command-line-register-lint-tool.rs)
-   [Tools registered in the crate root are recognized in submodules](https://github.com/rust-lang/rust/blob/master/src/test/ui/lint/known-tool-in-submodule)
-   [Unknown tools are reported](https://github.com/rust-lang/rust/blob/master/src/test/ui/lint/register-tool-lint.rs)

## Documentation

-   [PR to add `register_tool` documentation to the Rust Language Reference](https://github.com/rust-lang/reference/pull/1132)

## Unresolved Questions

A minor bug was [discovered](https://github.com/rust-lang/rust/issues/66079#issuecomment-925011317) and [fixed](https://github.com/rust-lang/rust/pull/89214) while this request was being prepared. To the best of my knowledge, there are no other outstanding issues.

Here is a [draft stabilization PR](https://github.com/rust-lang/rust/pull/92669).

**Comment by Aaron1011** [comment]
The attribute behavior (allowing `#[my_tool::my_attribute]`) is definitely not ready to be stabilized - this gives user code the ability to declare 'inert' attributes, which needs further discussion.

I think it would be okay to stabilize the lint behavior (suppressing 'unused lint' warnings for `#[deny(my_tool::my_lint)]`). However, we would need to split out that functionality into its own crate-level attribute (e.g. `#![register_lint_prefix]`).

cc @petrochenkov

**Comment by petrochenkov** [comment]
>this gives user code the ability to declare 'inert' attributes

But that's the whole point of tool attributes, both already stable ones (rustfmt and clippy prefixes), and custom ones declared with `register_tool`.
When writing the comment suggesting stabilization (https://github.com/rust-lang/rust/issues/66079#issuecomment-916984430), I meant inert attributes first of all.

**Comment by Aaron1011** [comment]
Ah, I didn't see you had originally suggested stabilization.

How does this feature interact with dependencies? If I (transitively) depend on a crate with a `#![register_tool]` attribute, will I 'import' all of the registrations? Or would I need to write `#![register_tool]` in every crate where I want to use it?

**Comment by petrochenkov** [comment]
It doesn't interact with dependencies, just allows the current crate to know about the tool.

**Comment by matklad** [comment]
> This probably needs to go through an RFC before stabilization.

Are we sure this doesn’t want an RFC? It feels like a rather cross-cutting issue and a new kind of public API.

In particular, I worry about cli-arg vs command line here. As a user, if I want to use, say, a custom `#[ide]` attribute, I’d probably want to declare that once in Cargo.toml, rather than in every test, binary and example.

That being said, an attribute is forward compatible with adding a CLI/Cargo.toml solution later, so it’s probably OK to just let this grow organically?

**Comment by petrochenkov** [comment]
>In particular, I worry about cli-arg vs command line here.

I initially implemented this as a command line option (https://github.com/rust-lang/rust/pull/57921), but switched to a crate-level attribute feedback from servo people.

FWIW, if you have an crate-level attribute, then you automatically have a command line option too at least on nightly, because of `-Z crate-attr` that allows to inject any attribute.

**Comment by joshtriplett** [comment]
What happens if there's a name conflict between a registered tool and a crate that might have an attribute macro in it? Does the macro still take effect? Does the name conflict get reported as an error?

**Comment by smoelius** [comment]
I did a quick test with the following code:
```rust
#![feature(register_tool)]
#![register_tool(log_derive)]

// use log_derive::logfn;

#[log_derive::logfn(Trace, fmt = "testing the num: {:?}")]
fn test_log(a: u8) -> String {
    (a * 2).to_string()
}

fn main() {
    env_logger::init();
    println!("{}", test_log(5));
}
```
As-is, the `logfn` macro takes effect.

However, if you uncomment the `use` declaration, you get:
```
error[E0659]: `log_derive` is ambiguous
 --> src/main.rs:4:5
  |
4 | use log_derive::logfn;
  |     ^^^^^^^^^^ ambiguous name
  |
  = note: ambiguous because of multiple potential import sources
  = note: `log_derive` could refer to a crate passed with `--extern`
  = help: use `::log_derive` to refer to this crate unambiguously
note: `log_derive` could also refer to the tool module defined here
 --> src/main.rs:2:18
  |
2 | #![register_tool(log_derive)]
  |                  ^^^^^^^^^^
```

**Comment by joshtriplett** [comment]
@smoelius Treating it as ambiguous seems like a great outcome here, since the tool could provide arbitrary attributes and the compiler doesn't know what it might provide. Thanks for testing!

**Comment by joshtriplett** [comment]
Nominating this for lang review. I think we'll want to FCP this, but I think it looks reasonable.

**Comment by smoelius** [comment]
Sorry, @joshtriplett, I should have been more clear. I think you spotted a bug in the implementation.

With the `use` declaration commented out, no ambiguity is reported.

In other words, I think perhaps this line ought to trigger such an error, though it doesn't currently:
```rust
#[log_derive::logfn(Trace, fmt = "testing the num: {:?}")]
```

**Comment by SimonSapin** [comment]
> I initially implemented this as a command line option (#57921), but switched to a crate-level attribute after feedback from servo people.

IIRC this feedback was against *rustc* command line options as the *only* interface, since they are not easy to set when using Cargo. See https://github.com/rust-lang/cargo/issues/5376

**Comment by petrochenkov** [comment]
@smoelius 
This is not a bug, this is usual shadowing rules, all prelude layers are walked in some order when looking up names, and extern crates are looked up first in that order.
See `fn visit_scopes` for the whole list of those layers (scopes).

**Comment by smoelius** [comment]
@petrochenkov Thank you for clarifying.

[`visit_scopes`](https://github.com/rust-lang/rust/blob/89b9f7b284aacc5f8613438b80e4dd7bdd10549e/compiler/rustc_resolve/src/lib.rs#L1724) - for anyone else confused by this.

**Comment by joshtriplett** [comment]
That seems potentially confusing, since rustc doesn't know what's valid and not within the namespace of a registered tool. Shouldn't this be treated as a name ambiguity, at least if the names are introduced in the same scope?

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting.

We felt like if this is going to be stabilized, it needs to have a clear specification, either in the form of an RFC, or at least more detailed documentation than the current PR to the reference.

In particular, the story for `register_tool(xyz)` and name conflicts with a dependency crate `xyz`, would be helpful to have clarification on.

We're also concnerned about the compatibility story here. Does this prevent us from adding new attributes, or new scoped attributes, without worrying about compatibility with existing third-party tools? Are we making a new stability commitment here?

Is every user of a tool expected to put `#![register_tool(thetool)]` in their code?

Also, as a bikeshed issue, folks felt like `register_tool` implied some higher degree of integration than it actually means here (in particular, that it might actually want to mean "please run this tool for me automatically" in the future). Something more like `ignore_...` might feel more accurate.

**Comment by smoelius** [comment]
Thanks for recounting, @joshtriplett.

**Comment by petrochenkov** [comment]
>In particular, the story for register_tool(xyz), and name conflicts with a dependency crate xyz

Meh, we have local variables shadowing each other and items, items shadowing dependency crates, almost anything potentially shadowing anything else in all namespaces, but when any new source of names is introduced there are always concerns "what if shadowing is bad, let's prohibit it in *this specific case*".

We shouldn't add such random artificial restrictions, IMO (and should relax the restrictions described in #56414 too), we just need to prominently document the order in which the names are searched following the design guidance described in `fn visit_scope` - more "user-controled" names and names from "more local scopes" are considered first.

**Comment by nikomatsakis** [comment]
I feel like this merits an RFC -- I'm feeling a bit confused about the overall intent for how this feature is to be used and anticipated workflows.

**Comment by joshtriplett** [comment]
To clarify, we don't necessarily need to prohibit name conflicts; rather, we need to document *what* happens in that case. If what should happen is "silently shadow", we should document that.

**Comment by bjorn3** [comment]
What about making `#![register_tool(...)]` "define" a module containing attribute macro definitions under every possible name and making `#![register_attr(...)]` "define" an attribute macro as far as name resolution is concerned. And then make it possible to publicly re-export said fake attribute macros. This way you could use `#![register_tool(xyz)]` and then `#[xyz::foo]` as consumer in case there is no crate `xyz` and use `#![register_attr(foo)] pub use foo;` in `xyz` in case such a crate exists and is used by the consumer that wants to use `#[xyz::foo]`.

**Comment by Mark-Simulacrum** [comment]
> I feel like this merits an RFC -- I'm feeling a bit confused about the overall intent for how this feature is to be used and anticipated workflows.

I agree with this -- it's not obvious to me that we want these in attributes, which are harder to e.g. manage across an entire workspace. I think Manish is right in https://github.com/rust-lang/rust/pull/57921#issuecomment-457855515 that -Z attributes *can* be harder to pass, particularly on a workspace basis today, but I'm not sure that's actually the right motivation for not doing things via the CLI, it just means that we need accompanying support in Cargo (or other build systems). That obviously requires its own design work.

I think that a design document laying out the details for how both end-users consuming tools and tool authors should think about these attributes would be useful in moving forward here.

**Comment by chorman0773** [comment]
I would definately like the attribute and lint controls separate.

For https://github.com/LightningCreations/lccc, I want the ability to support additional lints via its own namespace, however, the namespace is also used for attributes. Unlike, e.g. `#[rustfmt::skip]` these attributes aren't merely informative to the tool, but have actual semantics, and on lccc itself, they would be gated behind a feature, and should not be considered portable at all. In contrast, having additional lint controls is comparatively harmless. `#![allow(lccc::abi_incompatibility)]` should have no problems being just treated as a no-op by rustc, but `#[lccc::mangle_as = "std"]` would not be a good thing to have just silently accepted, as the person writing code using it may except that it has a semantic meaning where accepted - that is, it either affects the external name of the thing it's attached to, or it fails to compile (which indeed it does, it fails to compile outside of lccc or when the feature isn't enabled, or it affects the external name as prescribed by the ABI, such that the relevent name component(s) is set to `std`).

**Comment by xFrednet** [comment]
Hey lang team, could you maybe discuss my suggestion to reduce the E0710 error to a lint trigger? (See comment above)

@rustbot label +I-lang-nominated

**Comment by joshtriplett** [comment]
We talked about this in today's @rust-lang/lang meeting, and we still hold the position that this needs an RFC before stabilizing, to answer several of the questions here.

@xFrednet, I would have concerns about the approach of downgrading that from a hard error to a lint. It'd be OK to explore that in the "alternatives' section of the aforementioned RFC, though.

**Comment by joshtriplett** [comment]
Also, the questions in https://github.com/rust-lang/rust/issues/66079#issuecomment-1010266282 seem like they should still apply.

**Comment by joshtriplett** [comment]
@xFrednet, I would have concerns about the approach of downgrading that from a hard error to a lint. It'd be OK to explore that in the "alternatives' section of the aforementioned RFC, though.

**Comment by pnkfelix** [comment]
@xFrednet also, in case its not clear: the act of downgrading the existing hard error to a lint itself a stabilization action (since it would be part of the new stabilized surface of the language). Thus, the T-lang design team views such a path as best described in the context of an overall plan for stabilizing this feature as a whole.

**Comment by xFrednet** [comment]
Alright, thank you for the feedback! Then I'll retract the normination for now. I sadly won't have time in the coming months to work on this. If anyone else is interested, I'm happy to share my use case with [Marker](github.com/rust-marker/marker) and also help with the RFC.

**Comment by iddm** [comment]
Please pardon me for asking, but when is `register_tool` going to become stable? Are there any plans for that? Why is it still experimental?

**Comment by bjorn3** [comment]
https://github.com/rust-lang/rfcs/pull/3808 hasn't been accepted yet.

**Comment by Nadrieril** [comment]
> https://github.com/rust-lang/rfcs/pull/3808 hasn't been accepted yet.

Now it has! 🎉

**Comment by kornelski** [comment]
Any blockers to stabilzing this?

I'm implementing custom linters, and I'd love to use `register_lint_tool` for `#[allow(my_linter::custom_rule)]`. Requiring nightly for a bit of syntax sugar is an overkill, and falling back to some hack like `// @my_linter::custom_rule` is sad. 

I don't need the `extern crate tool` proc-macro-attr-like support, so to me stabilizing the easy lint name part without the rest would be okay too.

**Comment by jyn514** [comment]
@kornelski well, the main blocker is that the RFC hasn't been implemented yet. at a minimum we need the design questions in https://github.com/rust-lang/rust/pull/158146 to be resolved.

**Comment by kornelski** [comment]
I don't understand the use-case for advanced scoped name resolution for tools. I don't see why this is needed. The custom attributes aren't proc macros, so they don't do anything inside `rustc`. External tools that parse Rust source code may not have the same name resolution power as `rustc`, so supporting anything other than a fixed global string seems like an overkill to me. In my linter I'm going to have a hardcoded check for `my_linter::` attr via `syn`, regardless of what other complex naming paths the RFC allows. My linter won't be able to understand some `pub use my_linter as tricky; #[allow(self::tricky::custom_rule)]` anyway. 

I saw there was a rejected proposal to register tool names via CLI. That's a shame - that was simple, and would work with Cargo. I'd love Cargo to also understand custom tool names. I'd like some `[lints.my_linter] custom_rule = "allow"`, which Cargo will struggle with if there's no CLI flag for it. Cargo can't inject `#[register_lint_tool(my_linter)]`, but could inject `--register-lint-tool=my_linter`. To me the closest analogy to custom tools is custom `cfg()` names. These are already allowlisted simply via command line, without support for advanced scoped aliasing, renaming, namespacing, etc.

**Comment by jyn514** [comment]
let's not have involved discussions on the tracking issue please. i'll ping you on Zulip and if you still have questions after we've talked, feel free to post them here.
